File: src/main/java/com/pancisin/bookster/rest/controllers/AdministratorController.java
package com.pancisin.bookster.rest.controllers;

import com.pancisin.bookster.model.Administrator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.pancisin.bookster.repository.AdministratorRepository;

@RestController
@RequestMapping("/api/administrator/{pa_id}")
@PreAuthorize("hasPermission(#pa_id, 'administrator', 'update')")
public class AdministratorController {

	@Autowired
	private AdministratorRepository administratorRepository;

	@GetMapping
	public ResponseEntity<?> getAdministrator(@PathVariable Long pa_id) {
		return ResponseEntity.ok(administratorRepository.findOne(pa_id));
	}

	@PutMapping
	public ResponseEntity<?> putAdministrator(@PathVariable Long pa_id, @RequestBody Administrator pa) {
		Administrator stored = administratorRepository.findOne(pa_id);

		stored.setRole(pa.getRole());
		stored.setActive(pa.getActive());

		return ResponseEntity.ok(administratorRepository.save(stored));
	}

	@DeleteMapping
	public ResponseEntity<?> deleteAdministrator(@PathVariable Long pa_id) {
		administratorRepository.delete(pa_id);
		return ResponseEntity.ok("success");
	}
}

File: src/main/kotlin/com/pancisin/bookster/rest/controllers/v1/AdministratorController.kt
package com.pancisin.bookster.rest.controllers.v1

import com.pancisin.bookster.model.Administrator
import com.pancisin.bookster.repository.AdministratorRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/administrator/{id}")
@PreAuthorize("hasPermission(#id, 'administrator', 'update')")
class AdministratorController {

  @Autowired
  lateinit var administratorRepository: AdministratorRepository

  @GetMapping
  fun getAdministrator(@PathVariable id: Long) = ResponseEntity.ok(administratorRepository.findOne(id))

  @PutMapping
  fun putAdministrator(@PathVariable id: Long, @RequestBody administrator: Administrator): ResponseEntity<*> {
    return ResponseEntity.ok(administratorRepository.save(administratorRepository.findOne(id).copy(role = administrator.role, active = administrator.active)))
  }

  @DeleteMapping
  fun deleteAdministrator(@PathVariable id: Long): ResponseEntity<*> {
    administratorRepository.delete(id)
    return ResponseEntity.ok("success")
  }
}
